File: 7_reverse_int/7_reverse_int.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ARRAY(...) { __VA_ARGS__ }
#define ARRAY_LEN(ARR) (sizeof(ARR)/sizeof(*ARR))

int reverse(int x) {
  int ret = 0;
  int rem = 0;
  for(int i = 1; i < 10; i++) {
    rem = x % 10;
    x = x / 10; 
    ret = ret * 10 + rem;
    if(x == 0) break;
    printf("x = %d, rem = %d, ret = %d\n", x, rem, ret);
  }
  if(x) {
    if(
      ret < INT32_MIN/10 ||
      ret > INT32_MAX/10 ||
      (ret == INT32_MIN/10 && x < INT32_MIN%10) ||
      (ret == INT32_MAX/10 && x > INT32_MAX%10)
    ) return 0;
    ret = ret * 10 + x;
  }
  return ret;
}

void test_ri(int x, int expected) {
  static int case_count = 0;
  int ret = reverse(x);
  printf("Case%d: %d => %d (%d)\n", case_count++, x, ret, expected);
  if(ret != expected) {
    fprintf(stderr, "Error: %s:%d: expected %d, found %d\n", __FILE_NAME__, __LINE__, expected, ret);
    exit(1);
  }
}

int main(void) {
  puts("");

  test_ri(123, 321);
  test_ri(-123, -321);
  test_ri(120, 21);
  test_ri(1534236469, 0);
  test_ri(-1463847412, -2147483641);

      return 0;
}

File: 7_reverse_int/7_reverse_int.ts
function reverse(x: number): number {
    const str = [...x.toString()];
    let is_neg = false;
    if(str[0] == "-") {
        is_neg = true;
        str.shift();
    }
    const ret = (is_neg ? -1 : 1) * Number( str.reverse().reduce((a,b) => a + b, "") );
    if(ret <= -1*2**31 || ret >= 2**31 - 1) return 0;
    return ret;
};

let case_count = 0;

function test_ri(x: number, expected: number): void {
  const ret = reverse(x);
  console.log(`Case${case_count++}: ${x} => ${ret} (${expected})`);
  if(ret != expected) throw new Error(`WrongAnswer`);
}

console.log();

test_ri(123, 321);
test_ri(-123, -321);
test_ri(120, 21);
test_ri(1534236469, 0);

